File: backend/sales/demand.py
"""How full a night is, and how fast it is filling, for the line above the reserve button.

Urgency only works when it is true. Everything here is counted from completed orders, and the claim shown to a
visitor is chosen to match the numbers rather than the numbers being chosen to suit a claim:

  - "N reserved recently" is counted over an hour, six hours and a day, and the window shown is the one that
    holds the MOST bookings, ties going to the tighter one. All three are true statements about the same
    orders, so this picks the most informative of them rather than the most flattering: a window can never
    claim more than its own count, so "the last hour" can never describe something that took a day. Preferring
    the tightest instead used to throw bookings away, and across two nights it threw them away twice: four
    Tuesday seats and three Wednesday ones were being advertised as "5 in the last few hours" because
    Tuesday's own window had narrowed to the hour. The real number was 7.
  - nothing is shown at all below a floor, because "1 reserved in the last day" is an advert for an empty room.
  - the progress bar appears only once a third of the seats have gone. A bar showing 2 of 60 tells somebody the
    room will be empty, which is the opposite of the thing we are trying to say, and it would be true.
"""

from datetime import timedelta

from django.db.models import Sum
from django.utils import timezone

from catalog.models import TicketType
from sales.models import Order, OrderItem

# Tightest first, which is also the tie-break order. Each is (hours, the key the widget translates).
WINDOWS = ((1, 'in the last hour'), (6, 'in the last few hours'), (24, 'in the last day'))
# Below this a count is not social proof, it is an admission.
MIN_RECENT = 2
# A bar this empty discourages; above it, it persuades.
BAR_FROM = 0.33


def _sold_elsewhere(event):
    return sum(t.sold_elsewhere or 0 for t in event.ticket_types.filter(active=True, is_addon=False))


def _seats(rows):
    """Seats only. A round of drinks is not a seat and must never inflate how full the room looks."""
    total = rows.exclude(ticket_type__is_addon=True).aggregate(n=Sum('quantity'))['n']
    return total or 0


def demand(event):
    """{'capacity', 'taken', 'left', 'recent', 'recentWindow'} or None when the night tracks no capacity."""
    if event is None:
        return None
    capacity = sum(t.capacity for t in event.ticket_types.filter(active=True, is_addon=False)
                   if t.capacity is not None)
    if not capacity:
        return None

    completed = OrderItem.objects.filter(order__event=event, order__status=Order.COMPLETED)
    # Seats sold through a guest promoter are seats in this room. Leaving them out told a visitor the night was
    # emptier than it is, which is the opposite of what this line exists to do.
    taken = _seats(completed) + _sold_elsewhere(event)

    now = timezone.now()
    by_window = {label: _seats(completed.filter(order__completed_at__gte=now - timedelta(hours=hours)))
                 for hours, label in WINDOWS}
    recent, window = _best_window(by_window)

    return {
        'capacity': capacity,
        'taken': taken,
        'left': max(0, capacity - taken),
        # Only once the room is visibly filling; before that the honest picture is a discouraging one.
        'showBar': taken / capacity >= BAR_FROM,
        'recent': recent,
        'recentWindow': window,
        # All three, so a page showing several nights can add them up over the SAME window. Summing each
        # night's own chosen window and labelling the total with the widest of them undercounts, which is how
        # 7 bookings were being advertised as 5.
        'recentByWindow': by_window,
    }


def _best_window(by_window):
    """The window carrying the most bookings, ties to the tighter one, or nothing if none clears the floor."""
    best, label = 0, ''
    for _, candidate in WINDOWS:
        count = by_window.get(candidate, 0)
        if count >= MIN_RECENT and count > best:
            best, label = count, candidate
    return best, label


def demand_for(events):
    """The same thing for a whole listing page, in two queries rather than two per row.

    The events index shows a dozen nights. Calling demand() per row would be a dozen round trips to say the
    same thing, which is how a listing page quietly becomes the slowest page on the site.
    """
    events = [e for e in events if e is not None]
    if not events:
        return {}

    capacities, elsewhere = {}, {}
    for t in TicketType.objects.filter(event__in=events, active=True, is_addon=False):
        if t.capacity is not None:
            capacities[t.event_id] = capacities.get(t.event_id, 0) + t.capacity
        elsewhere[t.event_id] = elsewhere.get(t.event_id, 0) + (t.sold_elsewhere or 0)

    now = timezone.now()
    taken, recent = {}, {label: {} for _, label in WINDOWS}
    rows = (OrderItem.objects
            .filter(order__event__in=events, order__status=Order.COMPLETED)
            .exclude(ticket_type__is_addon=True)
            .values_list('order__event_id', 'quantity', 'order__completed_at'))
    for event_id, quantity, completed in rows:
        taken[event_id] = taken.get(event_id, 0) + quantity
        for hours, label in WINDOWS:
            if completed and completed >= now - timedelta(hours=hours):
                recent[label][event_id] = recent[label].get(event_id, 0) + quantity

    out = {}
    for event in events:
        capacity = capacities.get(event.id, 0)
        if not capacity:
            continue
        got = taken.get(event.id, 0) + elsewhere.get(event.id, 0)
        by_window = {label: recent[label].get(event.id, 0) for _, label in WINDOWS}
        count, window = _best_window(by_window)
        out[event.id] = {
            'capacity': capacity, 'taken': got, 'left': max(0, capacity - got),
            'showBar': got / capacity >= BAR_FROM, 'recent': count, 'recentWindow': window,
            'recentByWindow': by_window,
        }
    return out

File: src/lib/demand.ts
import type { KintanaPublicEvent } from "@kintana/sdk";

import type { Locale } from "../i18n/locale";
import { t } from "../i18n/ui";

/**
 * How full a night is, and how fast it is filling. The API sends this on every event (see
 * `backend/sales/demand.py`), and the checkout iframe already prints the same three sentences from the same
 * numbers. This is the site's copy of that, so a person sees one consistent story from the home page hero to
 * the reserve button.
 *
 * Urgency is only persuasive while it is true, so every rule that keeps it honest lives in the backend and is
 * reproduced here rather than loosened: the window is the tightest one that genuinely holds the bookings, a
 * count below two is not shown at all, and an empty room says nothing rather than something weak.
 */
export type EventDemand = {
  capacity: number;
  taken: number;
  left: number;
  showBar: boolean;
  recent: number;
  /** The English window key the backend chose; both languages map it through `WINDOW_KEYS`. */
  recentWindow: string;
  /** Every window's count, so several nights can be added up over the SAME window. */
  recentByWindow: Record<string, number>;
};

/** The SDK's event type is fixed and does not know about this field, so read it off the payload by hand. */
export function eventDemand(evt: KintanaPublicEvent): EventDemand | null {
  const raw = (evt as unknown as { demand?: unknown }).demand;
  if (!raw || typeof raw !== "object") return null;
  const d = raw as Partial<EventDemand>;
  if (!d.capacity || typeof d.capacity !== "number") return null;
  return {
    capacity: d.capacity,
    taken: Number(d.taken ?? 0),
    left: Number(d.left ?? 0),
    showBar: Boolean(d.showBar),
    recent: Number(d.recent ?? 0),
    recentWindow: String(d.recentWindow ?? ""),
    recentByWindow: (d.recentByWindow && typeof d.recentByWindow === "object" ? d.recentByWindow : {}) as Record<string, number>,
  };
}

/** Matches MIN_RECENT in `backend/sales/demand.py`. Below this a count is not social proof, it is an admission. */
const MIN_RECENT = 2;

const WINDOW_KEYS = {
  "in the last hour": "demand.window.hour",
  "in the last few hours": "demand.window.hours",
  "in the last day": "demand.window.day",
} as const;

export type DemandLine = {
  /**
   * The state of the room: sold out, nearly gone, over half, or a plain count. This is the limited-space fact,
   * and it belongs with the bar because it is what the bar is drawing.
   */
  room: string;
  /** Other people booking. Social proof, and empty below two, where a count is an admission rather than proof. */
  momentum: string;
  /**
   * The single best line, for somewhere that only has room for one (the home hero badge).
   *
   * 🚨 `room` and `momentum` are DIFFERENT KINDS of fact and must not compete for the same slot. They used to:
   * one ladder produced one sentence, momentum outranked half a room, and since almost every night here has
   * two or more recent bookings the room's own state was never said out loud anywhere. Measured 2026-09-25,
   * with Privilegio at 40 of 80 and the Tuesday open mic at 37 of 60, four surfaces printed "N people
   * reserved in the last day" and not one of them mentioned that half the seats had gone.
   */
  text: string;
  /** Genuinely nearly gone, so the line is worth colouring. */
  tight: boolean;
  /** 0-100, or null when the room is too empty for a bar to be anything but discouraging. */
  fill: number | null;
};

/**
 * The one sentence worth saying about this night, or null when there is nothing honest to say.
 *
 * Order matters: seats running out beats somebody else booking, which beats a plain count. Each rung is a
 * weaker argument than the one above it, and the bottom rung is silence.
 */
export function demandLine(evt: KintanaPublicEvent, locale: Locale): DemandLine | null {
  const d = eventDemand(evt);
  if (!d) return null;

  const gone = d.left === 0;
  // A fifth of the room, capped at ten. A flat "ten left" is nonsense on a small night: a room of three with
  // nothing sold has three left, and announcing "Only 3 seats left of 3" to an empty house is worse than
  // saying nothing, because it is a lie anybody can check by looking at the room.
  const nearlyGone = Math.min(10, Math.max(1, Math.floor(d.capacity * 0.2)));
  const tight = gone || (d.left > 0 && d.left <= nearlyGone);
  // Half a room is a quieter fact than a burst of bookings, but it stays true for longer, so it sits under
  // the recent line and above the plain count. Nothing is said about an empty room: that is not an argument
  // for coming, and a weak number reads as an admission.
  const half = d.capacity > 0 && d.taken / d.capacity >= 0.5;
  let room = "";
  if (gone) {
    room = t(locale, "demand.soldOut");
  } else if (tight) {
    room = t(locale, "demand.left", { count: String(d.left), total: String(d.capacity) });
  } else if (half) {
    room = t(locale, "demand.half");
  } else if (d.showBar) {
    room = t(locale, "demand.taken", { taken: String(d.taken), total: String(d.capacity) });
  }

  let momentum = "";
  if (!gone && d.recent >= MIN_RECENT && d.recentWindow in WINDOW_KEYS) {
    const when = t(locale, WINDOW_KEYS[d.recentWindow as keyof typeof WINDOW_KEYS]);
    momentum = t(locale, "demand.recent", { count: String(d.recent), window: when });
  }

  // For one slot only: running out beats somebody else booking, which beats a plain count. Unchanged, so the
  // hero badge reads exactly as it did.
  const text = gone || tight ? room : momentum || room;
  if (!text) return null;

  return { room, momentum, text, tight, fill: d.showBar ? Math.min(100, Math.round((d.taken / d.capacity) * 100)) : null };
}

/** Tightest first, matching WINDOWS in `backend/sales/demand.py`; also the tie-break order. */
const WINDOW_ORDER = ["in the last hour", "in the last few hours", "in the last day"];

/**
 * One sentence about the open mics as a whole, rather than one per night.
 *
 * The hero offers two nights side by side, and a line under each of them ("Tue · 2 people reserved in the last
 * few hours", "Wed · 3 people reserved in the last few hours") says the same thing twice and reads as filler.
 * Added together it is a bigger number, one line, and the same fact.
 *
 * The nights are added up over the SAME window, which is the whole reason the API sends every window's count.
 * Summing each night's own chosen window instead and labelling the total with the widest of them undercounts:
 * four Tuesday seats and three Wednesday ones came out as "5 in the last few hours" because Tuesday's own
 * window had narrowed to the hour. It was 7.
 */
export function combinedRecentLine(events: KintanaPublicEvent[], locale: Locale): string | null {
  const counted = events.map(eventDemand).filter((d): d is EventDemand => Boolean(d));
  if (!counted.length) return null;

  // Same rule as one night: the window holding the most bookings, ties to the tighter one.
  let best = 0;
  let window = "";
  for (const candidate of WINDOW_ORDER) {
    const total = counted.reduce((sum, d) => sum + (Number(d.recentByWindow[candidate]) || 0), 0);
    if (total >= MIN_RECENT && total > best) {
      best = total;
      window = candidate;
    }
  }
  if (!window) return null;

  const when = t(locale, WINDOW_KEYS[window as keyof typeof WINDOW_KEYS]);
  return t(locale, "demand.recent", { count: String(best), window: when });
}
